core/object/class_db.cpp

```diff
@@ -505,36 +505,85 @@ Object *ClassDB::_instantiate_internal(const StringName &p_class, bool p_require
 		ERR_FAIL_COND_V_MSG(ti->disabled, nullptr, "Class '" + String(p_class) + "' is disabled.");
 		ERR_FAIL_NULL_V_MSG(ti->creation_func, nullptr, "Class '" + String(p_class) + "' or its base class cannot be instantiated.");
 	}
+
 #ifdef TOOLS_ENABLED
 	if (ti->api == API_EDITOR && !Engine::get_singleton()->is_editor_hint()) {
 		ERR_PRINT("Class '" + String(p_class) + "' can only be instantiated by editor.");
 		return nullptr;
 	}
 #endif
-	if (ti->gdextension && ti->gdextension->create_instance) {
-		ObjectGDExtension *extension = ti->gdextension;
-#ifdef TOOLS_ENABLED
-		if (!p_require_real_class && ti->is_runtime && Engine::get_singleton()->is_editor_hint()) {
-			extension = get_placeholder_extension(ti->name);
-		}
-#endif
-		return (Object *)extension->create_instance(extension->class_userdata);
-	} else {
+
 #ifdef TOOLS_ENABLED
-		if (!p_require_real_class && ti->is_runtime && Engine::get_singleton()->is_editor_hint()) {
-			if (!ti->inherits_ptr || !ti->inherits_ptr->creation_func) {
-				ERR_PRINT_ONCE(vformat("Cannot make a placeholder instance of runtime class %s because its parent cannot be constructed.", ti->name));
-			} else {
-				ObjectGDExtension *extension = get_placeholder_extension(ti->name);
-				return (Object *)extension->create_instance(extension->class_userdata);
+	// Try to create placeholder.
+	if (!p_require_real_class && ti->is_runtime && Engine::get_singleton()->is_editor_hint()) {
+		bool can_create_placeholder = false;
+		if (ti->gdextension) {
+			if (ti->gdextension->create_instance2) {
+				can_create_placeholder = true;
+			}
+#ifndef DISABLE_DEPRECATED
+			else if (ti->gdextension->create_instance) {
+				can_create_placeholder = true;
+			}
+#endif // DISABLE_DEPRECATED
+		} else if (!ti->inherits_ptr || !ti->inherits_ptr->creation_func) {
+			// Only print error message once for each classes.
+			static LocalVector<StringName> err_printed_classes;
+			if (err_printed_classes.find(ti->name) >= 0) {
+				err_printed_classes.push_back(ti->name);
+				ERR_PRINT(vformat("Cannot make a placeholder instance of runtime class %s because its parent cannot be constructed.", ti->name));
```

Comment:
Why is this doing something custom, rather than `ERR_PRINT_ONCE()`?

Reply:
If use `ERR_PRINT_ONCE()` In this situation, it would print once if there have multiple classes encount this error.

Reply:
Ah, ok, that makes sense. However, I wonder if ensuring this is worth the extra code? Maybe we should just do plain `ERR_PRINT()` here without trying to avoid repeatedly printing this message? So:

```suggestion
			ERR_PRINT(vformat("Cannot make a placeholder instance of runtime class %s because its parent cannot be constructed.", ti->name));
```

Reply:
I‘m not sure, I just obey the previous implementation (`ERR_PRINT_ONCE()`). Maybe there has a situation that result in printing this error continuously?

Reply:
I'm the one who added the `ERR_PRINT_ONCE()`. :-) I never actually saw it get printed continuously, I just figured it wasn't very useful to see it more than once, but I didn't really understand how `ERR_PRINT_ONCE()` works, and didn't realize it would prevent seeing the message for the different classes.

Personally, I think it'd be best to just switch to plain `ERR_PRINT()` without the extra code, and we can adjust it later if it turns out that this produces error spam in practice.